src/controllers/sharedExecutionPlanUtils.ts

```diff
@@ -0,0 +1,184 @@
+    }
+
+    // Show a save dialog to the user
+    const saveUri = await vscode.window.showSaveDialog({
+        defaultUri: filename,
+        filters: {
+            "SQL Plan Files": ["sqlplan"],
+        },
+    });
+
+    if (saveUri) {
+        // Write the content to the new file
+        void vscode.workspace.fs.writeFile(
+            saveUri,
+            Buffer.from(payload.sqlPlanContent),
+        );
+    }
+
+    return state;
+}
+
+export async function showPlanXml(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["showPlanXml"],
+) {
+    const planXmlDoc = await vscode.workspace.openTextDocument({
+        content: formatXml(payload.sqlPlanContent),
+        language: "xml",
+    });
+
+    void vscode.window.showTextDocument(planXmlDoc);
+
+    return state;
+}
+
+export async function showQuery(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["showQuery"],
+    untitledSqlDocumentService: UntitledSqlDocumentService,
+) {
+    void untitledSqlDocumentService.newQuery(payload.query);
+
+    return state;
+}
+
+export async function updateTotalCost(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    payload: ExecutionPlanReducers["updateTotalCost"],
+) {
+    return {
+        ...state,
+        executionPlanState: {
+            ...state.executionPlanState,
+            totalCost: (state.executionPlanState.totalCost +=
+                payload.addedCost),
+        },
+    };
+}
+
+export async function createExecutionPlanGraphs(
+    state: QueryResultWebviewState | ExecutionPlanWebviewState,
+    executionPlanService: ExecutionPlanService,
+    xmlPlans: string[],
+) {
+    let newState = {
+        ...state.executionPlanState,
+    };
+    const startTime = performance.now(); // timer for telemetry
+    for (const plan of xmlPlans) {
+        const planFile: ExecutionPlanGraphInfo = {
+            graphFileContent: plan,
+            graphFileType: ".sqlplan",
+        };
+        try {
+            newState.executionPlanGraphs = newState.executionPlanGraphs.concat(
+                (await executionPlanService.getExecutionPlan(planFile)).graphs,
+            );
+
+            newState.loadState = ApiStatus.Loaded;
+
+            sendActionEvent(
+                TelemetryViews.QueryPlan,
+                TelemetryActions.OpenQueryPlan,
+                {},
+                {
+                    numberOfPlans:
+                        state.executionPlanState.executionPlanGraphs.length,
+                    LoadTimeInMs: performance.now() - startTime,
+                },
+            );
+        } catch (e) {
+            newState.loadState = ApiStatus.Error;
+            newState.errorMessage = e.toString();
+            sendErrorEvent(
+                TelemetryViews.QueryPlan,
+                TelemetryActions.OpenQueryPlan,
+                e,
+                true, // includeErrorMessage

```

Comment:
Can you confirm that this error message won't have any PII in it?  Otherwise, you may need to try to extract something useful out of it for a property and set this to `false`.

Reply:
Is there a need to log these error? What will we do once we get these errors. It is rare that execution plan parser has errors in it, most likely it will be malformed plan file.

Reply:
@Benjin , @laurennat

Reply:
I think we should attempt to detect if it's a malformed XML error and _not_ call `sendErrorEvent()` in that case.  Seems silly to have an opportunity to detect unknown errors here and not emit an error event if they aren't user errors.

Reply:
i was looking at this more and I think the only errors actually thrown after the refactor are the malformed plans. I'll just remove the send error event